### PR TITLE
Add --csv-file option for 'nucypher stake events' CLI to allow events to be output to CSV

### LINKED ISSUE #2547: Allow CSV output file option for `nucypher stake events` CLI output
For easier accounting of staker information, look into outputting the results of events into a CSV output file.

### PR DESCRIPTION
**Pull request by derekpierre**
**Type of PR:**
- [ ] Bugfix
- [ ] Feature
- [ ] Documentation
- [X] Other

**Required reviews:** 
- [ ] 1
- [X] 2
- [ ] 3

**What this does:**
> High-level idea of the changes introduced in this PR. 
> List relevant API changes (if any), as well as related PRs and issues.

Allows event information to be output to a csv file - this allows for easier accounting of staker behaviour.

**Issues fixed/closed:**
> - Fixes #...

Fixes #2547 
Follow-up issue: #2561 

**Why it's needed:**
> Explain how this PR fits in the greater context of the NuCypher Network.
> E.g., if this PR address a `nucypher/productdev` issue, let reviewers know!

Makes accounting of tokens easier for the staker.

**Notes for reviewers:**
> What should reviewers focus on? 
> Is there a particular commit/function/section of your PR that requires more attention from reviewers?

CSV functionality.

For a test, run 
`nucypher stake events --staking-address <staking address> --provider <provider> --event-name <event name> --csv-file <filepath>`

OR 

`nucypher stake events --staking-address <staking address> --provider <provider> --event-name <event name> --csv`

Open the csv file using an appropriate application.

### CODE REVIEW

File: nucypher/cli/commands/stake.py [hunk 1]
```diff
@@ -1308,24 +1309,43 @@ def events(general_config, staker_options, config_file, event_name):
...
+    else:
+        title = f" {contract_name} Events ".center(40, "-")
+        emitter.echo(f"\n{title}\n", bold=True, color='green')
+        emitter.echo(f"{event_type.event_name}:", bold=True, color='yellow')
+        entries = event_type.getLogs(fromBlock=0, toBlock='latest', argument_filters=argument_filters)
         for event_record in entries:
             emitter.echo(f"  - {EventRecord(event_record)}")
 
 
+def _write_events_csv_file(csv_file, event_args_filter, event_type):
```
Comment (KPrasch) [inline nucypher/cli/commands/stake.py:1335]: Consider relocating this method to a helper module.

File: nucypher/cli/commands/stake.py [hunk 2]
```diff
@@ -130,6 +129,7 @@
 from nucypher.config.characters import StakeHolderConfiguration
 from nucypher.utilities.gas_strategies import construct_fixed_price_gas_strategy
 
+option_csv_file = click.option('--csv-file', help="Filepath for CSV output", type=click.Path(dir_okay=False))
```
Comment (KPrasch) [inline nucypher/cli/commands/stake.py:132]: Consider a two flag approach, one using the a default, generated filepath, and the other to customize it.

**Review by KPrasch — COMMENTED** [review]
:+1: I was able to successfully create a file!  

A few suggestions:
- Public-facing documentation and examples
- Include the event name as a column in the CSV
- Prevent overwriting an existing file
- Use a default filename/structure (perhaps with timestamp)

Some ideas that will improve the experience further but are certainly outside the scope of this PR (perhaps in #2528):
- Allow for this report to be generated without a signer, using checksum
- Accept several addresses at once, Generate a compiled report

File: nucypher/cli/commands/stake.py (hunk 1 above)
Comment (derekpierre) [reply nucypher/cli/commands/stake.py]: ✔️

File: nucypher/cli/commands/stake.py (hunk 2 above)
Comment (derekpierre) [reply nucypher/cli/commands/stake.py]: ✔️

**Comment by derekpierre** [comment]
Still need to write some documentation.

**Comment by derekpierre** [comment]
> A few suggestions:
> 
> Public-facing documentation and examples
> Include the event name as a column in the CSV
> Prevent overwriting an existing file
> Use a default filename/structure (perhaps with timestamp)

✔️ - Fixed 

> Allow for this report to be generated without a signer, using checksum
> Accept several addresses at once, Generate a compiled report

@KPrasch filed #2561 .

File: nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol [hunk 3]
```diff
@@ -109,25 +109,113 @@ contract StakingEscrow is Issuer, IERC900History {
...
+    * @notice Signals that two sub-stakes were merged
+    * @param staker Staker address
+    * @param value1 Value of first sub-stake (in NuNits)
+    * @param value2 Value of second sub-stake (in NuNits)
+    * @param lastPeriod Final locked period of merged sub-stake
+    */
     event Merged(address indexed staker, uint256 value1, uint256 value2, uint16 lastPeriod);
+
+    /**
+    * @notice Signals that a sub-stakes was prolonged
```
Comment (vzotova) [inline nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol:155]:
```suggestion
    * @notice Signals that a sub-stake was prolonged
```

File: nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol [hunk 4]
```diff
@@ -109,25 +109,113 @@ contract StakingEscrow is Issuer, IERC900History {
...
+    * @param staker Staker address
+    * @param value Value of sub-stake
+    * @param lastPeriod Final locked period of sub-stake
```
Comment (vzotova) [inline nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol:158]:
```suggestion
    * @param lastPeriod Final locked period of old sub-stake
```

File: nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol [hunk 5]
```diff
@@ -109,25 +109,113 @@ contract StakingEscrow is Issuer, IERC900History {
...
+    * @param periods Number of periods sub-stake was extended
+    */
     event Prolonged(address indexed staker, uint256 value, uint16 lastPeriod, uint16 periods);
+
+    /**
+    * @notice Signals that tokens were withdrawn to the staker
```
Comment (vzotova) [inline nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol:164]: Small nitpick: in some events `staker` starts from lowercase letter in other places - from capital

File: nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol [hunk 6]
```diff
@@ -109,25 +109,113 @@ contract StakingEscrow is Issuer, IERC900History {
...
+
+    /**
+    * @notice Signals that the winddown parameter was activated/deactivated
+    * @param staker Staker address
+    * @param windDown Updated parameter value
+    */
     event WindDownSet(address indexed staker, bool windDown);
+
+    // undocumented
+    event WorkMeasurementSet(address indexed staker, bool measureWork);
```
Comment (vzotova) [inline nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol:218]: Maybe say something like internal event? (you can use `///` for one-liner doc)

File: nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol [hunk 7]
```diff
@@ -109,25 +109,113 @@ contract StakingEscrow is Issuer, IERC900History {
...
     event SnapshotSet(address indexed staker, bool snapshotsEnabled);
```
Comment (vzotova) [inline nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol:219]: This event same as winddown and restake events

File: nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol (hunk 5 above)
Comment (derekpierre) [reply nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol:164]: ✔️

File: nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol (hunk 6 above)
Comment (derekpierre) [reply nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol]: ✔️

File: nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol (hunk 7 above)
Comment (derekpierre) [reply nucypher/blockchain/eth/sol/source/contracts/StakingEscrow.sol:222]: ✔️

**Comment by KPrasch** [comment]
Related to #2559

File: nucypher/cli/commands/stake.py [hunk 8]
```diff
@@ -1276,35 +1282,56 @@ def preallocation(general_config: GroupGeneralConfig,
...
+def events(general_config, staker_options, config_file, event_name, csv, csv_file):
+    """View blockchain events associated with a staker"""
 
     # Setup
     emitter = setup_emitter(general_config)
-    STAKEHOLDER = staker_options.create_character(emitter, config_file)
 
+    if not event_name:
+        raise click.BadOptionUsage(message='You must specify an event name with --event-name')
+        # TODO: Doesn't work for the moment
```
Comment (KPrasch) [inline nucypher/cli/commands/stake.py:1296]: Track this TODO with an issue number?

**Review by KPrasch — APPROVED** [review]
Looking Great @derekpierre , Thanks!

File: nucypher/cli/commands/stake.py (hunk 8 above)
Comment (derekpierre) [reply nucypher/cli/commands/stake.py:1296]: @cygnusv is this TODO still applicable? It was added via https://github.com/nucypher/nucypher/commits/a51f675fdacecb78a0f91102cd0dc28a4ae18aef.

**Review by vzotova — APPROVED** [review]

**Pull request merged**